### PR TITLE
INDDEX reports take 2

### PR DESCRIPTION
**Pull request by esoergel**
https://docs.google.com/document/d/1fy7_JYhZwy4z8VA8ikygXSXTTSHNqaTFUWaj1AHcVD0/edit

##### SUMMARY
Picking up from https://github.com/dimagi/commcare-hq/pull/26885

This is where I'm at now with the INDDEX custom reports.  It's mostly new code, which will eventually replace the old code.  I'm developing it in separate and leaving the old code around for reference until it's ready.  It'll see some more changes, of course, but this should give an idea of where it's headed.

There are three main components here, plus tests:

* A UCR adapter class which only returns data from the UCR, unmodified
* A series of fixture adapters which make explicit the schema of the fixtures and provide easy access
* A primary report data generator, which pulls in data from the above two sources and produces a list of rows.

Those components are wired together, but not into anything else, so this shouldn't yet appear in the UI.

Review by commit.

### CODE REVIEW

File: custom/inddex/food.py [hunk 1]
```diff
@@ -1,175 +1,188 @@
...
-    'protein_g_per_100g',
-    'protein_g',
-    'conv_factor_gap_code',
-    'conv_factor_gap_desc',
-    'fct_gap_code',
-    'fct_gap_desc',
+MISSING = ''
+
+IN_UCR = 'in_ucr'
+IS_RECALL_META = 'is_recall_meta'
```
Comment (esoergel) [inline custom/inddex/food.py]: My aim here is to create a series of tags to document a bit about how each indicator is used and where it comes from (and to hopefully avoid a series of lists like "indicators, indicators_in_ucr, recall_meta_indicators, indicators_in_food_fixture...")

File: custom/inddex/fixtures.py [hunk 2]
```diff
@@ -34,4 +33,6 @@ def get_recipes(self):
         for item in iter_fixture_items_for_data_type(self.domain, recipes_data_type.get_id):
             item_dict = {field_name: field_list.field_list[0].field_value
                          for field_name, field_list in item.fields.items()}
-            yield RecipeIngredient(**item_dict)
+            # The fixture contains duplicate entries for each language.
+            if item_dict.pop('iso_code') == 'en':
```
Comment (kaapstorm) [inline custom/inddex/fixtures.py]: Is it safe to assume that it is always English that we want?

File: custom/inddex/example_data/data.py [hunk 3]
```diff
@@ -68,7 +68,8 @@ def _get_importer_config(case_type, headers, user_id):
 
 
 def get_expected_report():
-    return _read_csv('expected_result.csv')
+    headers, rows = _read_csv('expected_result.csv')
+    return [dict(zip(headers, row)) for row in rows]
```
Comment (kaapstorm) [inline custom/inddex/example_data/data.py]: :black_heart: So much more manageable this way.

**Review by kaapstorm — APPROVED** [review]
Great tests. I like where this is going.

File: custom/inddex/ucr_data.py [hunk 4]
```diff
@@ -3,28 +3,22 @@
...
-        'nsr_conv_option_desc_post_cooking', 'nsr_measurement_amount_post_cooking', 'nsr_same_conv_method',
-        'respondent_id',
-    ]
-    columns = [DatabaseColumn(col_id, SimpleColumn(col_id)) for col_id in UCR_COLUMN_IDS]
+
+    @property
+    def columns(self):
+        column_ids = [
+            'doc_id', 'inserted_at', 'nsr_same_conv_method',  # TODO remove this one?
+        ] + [i.slug for i in INDICATORS if i.in_ucr]
```
Comment (kaapstorm) [inline custom/inddex/ucr_data.py]: :+1:

File: custom/inddex/fixtures.py (hunk 2 above)
Comment (esoergel) [reply custom/inddex/fixtures.py]: So far, I think so.

File: custom/up_nrhm/tests/__init__.py [hunk 5]
```diff
@@ -45,11 +43,8 @@ def setUpModule():
     _call_center_domain_mock.stop()
 
 
+@db_tests_only
 def tearDownModule():
-    if isinstance(Domain.get_db(), Mock):
-        # needed to skip setUp for javascript tests thread on Travis
-        return
-
     _call_center_domain_mock = mock.patch(
```
Comment (esoergel) [inline custom/up_nrhm/tests/__init__.py]: I also notice that the rest of these setup/teardown functions look suspiciously similar.  If I look too long at this I'd probably get sucked in to a refactor

:see_no_evil:

File: corehq/util/test_utils.py [hunk 6]
```diff
@@ -714,3 +715,13 @@ def print_sql_chronologically(self, with_traceback=False, width=150):
             print('\n{}'.format(indent('\n'.join(wrap(out, width)), '\t')))
             if with_traceback:
                 print('\n{}'.format(indent(''.join(query['traceback']), '\t\t')))
+
+
+def db_tests_only(fn):
+    """Skip setUpModule and tearDownModule in non-DB contexts"""
```
Comment (sravfeyn) [inline corehq/util/test_utils.py]: nit: This comment indicates the context in which this method is used, just a bit confusing.

File: custom/champ/tests/__init__.py [hunk 7]
```diff
@@ -1,23 +1,21 @@
...
 from corehq.apps.userreports.models import StaticDataSourceConfiguration
 from corehq.apps.userreports.util import get_indicator_adapter, get_table_name
-from corehq.sql_db.connections import connection_manager, UCR_ENGINE_ID
+from corehq.sql_db.connections import UCR_ENGINE_ID, connection_manager
+from corehq.util.test_utils import db_tests_only
 
 
+@db_tests_only
 def setUpModule():
-    if isinstance(Domain.get_db(), Mock):
```
Comment (sravfeyn) [inline custom/champ/tests/__init__.py]: Nice cleanup!

File: custom/inddex/food.py [hunk 8]
```diff
@@ -0,0 +1,212 @@
+from .fixtures import FixtureAccessor
+
+MISSING = ''
+
+IN_UCR = 'in_ucr'
+IN_FOOD_FIXTURE = 'in_food_fixture'
+IS_RECALL_META = 'is_recall_meta'
+
+
+class I:
```
Comment (sravfeyn) [inline custom/inddex/food.py]: I for Indicator?

**Review by sravfeyn — APPROVED** [review]

File: corehq/util/test_utils.py (hunk 6 above)
Comment (esoergel) [reply corehq/util/test_utils.py]: I intended that to be a description of what this does, got any suggestions on how to word it better?

File: custom/inddex/food.py (hunk 8 above)
Comment (esoergel) [reply custom/inddex/food.py]: Yeah - I know single-letter class names are generally frowned upon, but I figured in this case things would be _more_ readable than repeating the full word 100 times.  Could also define the class as `Indicator` then immediately alias as `I`, or use plain tuples in the big list and generate the classes from them later.  I dunno, I think this way is fine.

**Pull request merged**